File: treversal/cpp-solution/zig zag treversal.cpp
// time: O(n) and stack: O(n)
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> result;
        if (root == nullptr){
            return result;
        }
        vector<TreeNode*> even_level;
        vector<TreeNode*> odd_level;
        odd_level.push_back(root);
        while (even_level.size() > 0 || odd_level.size() > 0){
            vector<int> temp;
            while(odd_level.size() > 0){
                TreeNode * current = odd_level.back();
                odd_level.pop_back();
                temp.push_back(current->val);
                // in the next level(even level) we need to treverse from right to left
                // so in this level if we push left node first, at the end we will end up 
                // with right node on top of the stack
                if (current->left) even_level.push_back(current->left);
                if (current->right) even_level.push_back(current->right);
            }
            if (temp.size() > 0){
                result.push_back(temp);
            }
            temp = {};
            while(even_level.size() > 0){
                TreeNode * current = even_level.back();
                even_level.pop_back();
                temp.push_back(current->val);
                if (current->right) odd_level.push_back(current->right);
                if (current->left) odd_level.push_back(current->left);
            }
            if (temp.size() > 0){
                result.push_back(temp);
            }
        }
        return result;
    }
};

File: treversal/python solution/zig zag treversal.py
# time: O(n) and space: O(n)
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        result = []
        if root is None:
            return result
        evenLevel= []
        oddLevel = []
        oddLevel.append(root)
        while evenLevel or oddLevel:
            currentLevel = []
            current = None
            # in the next level(even level) we need to treverse from right to left
            # so in this level if we push left node first, at the end we will end up 
            # with right node on top of the stack
            while oddLevel:
                current = oddLevel.pop()
                currentLevel.append(current.val)
                if current.left:
                    evenLevel.append(current.left)
                if current.right:
                    evenLevel.append(current.right)
            if currentLevel:
                result.append(currentLevel)
            currentLevel = []
            while evenLevel:
                current = evenLevel.pop()
                currentLevel.append(current.val)
                if current.right:
                    oddLevel.append(current.right)
                if current.left:
                    oddLevel.append(current.left)
            if currentLevel:
                result.append(currentLevel)
        return result
                
